File: packages/python/stagehand_social/core/session.py
"""Session manager — mirror of packages/typescript/src/core/session.ts."""

from __future__ import annotations

import json
import os
from dataclasses import asdict
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from browserbase import Browserbase

from stagehand_social.core.errors import SessionError
from stagehand_social.core.types import (
    GenericTarget,
    PlatformTarget,
    SessionRef,
    TargetKey,
)

_DEFAULT_SESSION_DIR = Path.home() / ".stagehand-social"
_SESSIONS_FILE = "sessions.json"


def _target_key_id(target: TargetKey) -> str:
    if isinstance(target, PlatformTarget):
        return target.platform
    return f"generic:{target.site_id}"


def _ref_key(target: TargetKey, account: str) -> str:
    return f"{_target_key_id(target)}::{account}"


def _target_to_dict(target: TargetKey) -> dict:
    return asdict(target)


def _target_from_dict(d: dict) -> TargetKey:
    if d.get("kind") == "platform":
        return PlatformTarget(platform=d["platform"])
    return GenericTarget(site_id=d["site_id"])


class SessionManager:
    """One Browserbase context per (target, account); persisted on disk."""

    def __init__(
        self,
        api_key: str,
        project_id: str,
        session_dir: Optional[str] = None,
    ):
        self.session_dir: Path = Path(session_dir) if session_dir else _DEFAULT_SESSION_DIR
        self.bb = Browserbase(api_key=api_key)
        self.project_id = project_id
        self._cache: dict[str, SessionRef] = {}
        self._loaded = False

    def _ensure_loaded(self) -> None:
        if self._loaded:
            return
        try:
            self.session_dir.mkdir(parents=True, exist_ok=True)
            file_path = self.session_dir / _SESSIONS_FILE
            if file_path.exists():
                raw = json.loads(file_path.read_text())
                self._cache = {
                    k: SessionRef(
                        target=_target_from_dict(v["target"]),
                        account=v["account"],
                        context_id=v["context_id"],
                        last_verified_at=v.get("last_verified_at"),
                    )
                    for k, v in raw.items()
                }
        except Exception as e:
            raise SessionError(f"Failed to load session cache from {self.session_dir}", e) from e
        self._loaded = True

    def _persist(self) -> None:
        file_path = self.session_dir / _SESSIONS_FILE
        serialized = {
            k: {
                "target": _target_to_dict(v.target),
                "account": v.account,
                "context_id": v.context_id,
                "last_verified_at": v.last_verified_at,
            }
            for k, v in self._cache.items()
        }
        file_path.write_text(json.dumps(serialized, indent=2))

    def get(self, target: TargetKey, account: str) -> Optional[SessionRef]:
        self._ensure_loaded()
        return self._cache.get(_ref_key(target, account))

    def get_or_create_context(self, target: TargetKey, account: str) -> str:
        """Return an existing contextId for (target, account), or create one."""
        self._ensure_loaded()
        existing = self._cache.get(_ref_key(target, account))
        if existing is not None:
            return existing.context_id

        ctx = self.bb.contexts.create(project_id=self.project_id)
        ref = SessionRef(target=target, account=account, context_id=ctx.id)
        self._cache[_ref_key(target, account)] = ref
        self._persist()
        return ctx.id

    def mark_verified(self, target: TargetKey, account: str) -> None:
        self._ensure_loaded()
        ref = self._cache.get(_ref_key(target, account))
        if ref is None:
            return
        ref.last_verified_at = datetime.now(tz=timezone.utc).isoformat()
        self._persist()

    def forget(self, target: TargetKey, account: str) -> None:
        self._ensure_loaded()
        self._cache.pop(_ref_key(target, account), None)
        self._persist()

    def list(self) -> list[SessionRef]:
        self._ensure_loaded()
        return list(self._cache.values())

    def build_session_params(self, context_id: str) -> dict:
        """Stagehand-compatible session params."""
        return {
            "project_id": self.project_id,
            "browser_settings": {
                "context": {"id": context_id, "persist": True},
            },
        }

File: packages/typescript/src/core/session.ts
/**
 * Session manager — tracks one Browserbase context per (target, account)
 * and persists references on disk so logins survive across runs.
 *
 * On disk format (JSON):
 *   ~/.stagehand-social/sessions.json
 *   {
 *     "instagram::myaccount": {
 *       "target": { "kind": "platform", "platform": "instagram" },
 *       "account": "myaccount",
 *       "contextId": "ctx_abc123",
 *       "lastVerifiedAt": "2026-05-02T..."
 *     }
 *   }
 */

import { promises as fs } from "node:fs";
import path from "node:path";
import os from "node:os";
import Browserbase from "@browserbasehq/sdk";
import type { SessionRef, TargetKey } from "./types.js";
import { SessionError } from "./errors.js";

const DEFAULT_SESSION_DIR = path.join(os.homedir(), ".stagehand-social");
const SESSIONS_FILE = "sessions.json";

function targetKeyId(target: TargetKey): string {
  return target.kind === "platform"
    ? target.platform
    : `generic:${target.siteId}`;
}

function refKey(target: TargetKey, account: string): string {
  return `${targetKeyId(target)}::${account}`;
}

export class SessionManager {
  private cache: Record<string, SessionRef> = {};
  private loaded = false;
  private readonly sessionDir: string;
  private readonly bb: Browserbase;
  private readonly projectId: string;

  constructor(opts: {
    apiKey: string;
    projectId: string;
    sessionDir?: string;
  }) {
    this.sessionDir = opts.sessionDir ?? DEFAULT_SESSION_DIR;
    this.bb = new Browserbase({ apiKey: opts.apiKey });
    this.projectId = opts.projectId;
  }

  private async ensureLoaded(): Promise<void> {
    if (this.loaded) return;
    try {
      await fs.mkdir(this.sessionDir, { recursive: true });
      const filePath = path.join(this.sessionDir, SESSIONS_FILE);
      const raw = await fs.readFile(filePath, "utf-8").catch(() => "{}");
      this.cache = JSON.parse(raw) as Record<string, SessionRef>;
    } catch (err) {
      throw new SessionError(
        `Failed to load session cache from ${this.sessionDir}`,
        err,
      );
    }
    this.loaded = true;
  }

  private async persist(): Promise<void> {
    const filePath = path.join(this.sessionDir, SESSIONS_FILE);
    await fs.writeFile(filePath, JSON.stringify(this.cache, null, 2), "utf-8");
  }

  /** Look up an existing session ref. Returns undefined if none. */
  async get(target: TargetKey, account: string): Promise<SessionRef | undefined> {
    await this.ensureLoaded();
    return this.cache[refKey(target, account)];
  }

  /**
   * Get an existing context for (target, account), or create a fresh one.
   * Returns the contextId — pass this to Stagehand's `browserbaseSessionCreateParams`.
   */
  async getOrCreateContext(
    target: TargetKey,
    account: string,
  ): Promise<string> {
    await this.ensureLoaded();
    const existing = this.cache[refKey(target, account)];
    if (existing) return existing.contextId;

    const ctx = await this.bb.contexts.create({ projectId: this.projectId });
    const ref: SessionRef = {
      target,
      account,
      contextId: ctx.id,
    };
    this.cache[refKey(target, account)] = ref;
    await this.persist();
    return ctx.id;
  }

  /** Mark a session as verified-logged-in right now. */
  async markVerified(target: TargetKey, account: string): Promise<void> {
    await this.ensureLoaded();
    const key = refKey(target, account);
    const ref = this.cache[key];
    if (!ref) return;
    ref.lastVerifiedAt = new Date().toISOString();
    await this.persist();
  }

  /** Forget a session (e.g., after logout). Does NOT delete the Browserbase context. */
  async forget(target: TargetKey, account: string): Promise<void> {
    await this.ensureLoaded();
    delete this.cache[refKey(target, account)];
    await this.persist();
  }

  /** List all known sessions. */
  async list(): Promise<SessionRef[]> {
    await this.ensureLoaded();
    return Object.values(this.cache);
  }

  /** Build the Browserbase session params Stagehand expects. */
  buildSessionParams(contextId: string): {
    projectId: string;
    browserSettings: { context: { id: string; persist: boolean } };
  } {
    return {
      projectId: this.projectId,
      browserSettings: {
        context: { id: contextId, persist: true },
      },
    };
  }
}
